File: core/utils/workflow/self_review.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
workflow self-review helper。

聚合 spec / plan 自检与文档契约检查，复用 traceability 与 doc_contracts。
"""

from __future__ import annotations

import json
import sys
from typing import Any, Dict, List, Optional

from doc_contracts import validate_workflow_doc_contracts
from task_parser import parse_tasks_v2
from traceability import (
    RequirementTrace,
    validate_plan_traceability,
    validate_spec_traceability,
)


def build_requirements(items: Optional[List[Dict[str, Any]]]) -> List[RequirementTrace]:
    return [
        RequirementTrace(
            id=item.get("id", ""),
            summary=item.get("summary", ""),
            scope_status=item.get("scope_status", "in_scope"),
            constraints=item.get("constraints") or [],
            owner=item.get("owner", "shared"),
            exclusion_reason=item.get("exclusion_reason"),
        )
        for item in (items or [])
        if item.get("id")
    ]


def run_spec_self_review(
    requirements: List[Dict[str, Any]],
    spec_content: str,
    ux_artifact: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    requirement_records = build_requirements(requirements)
    result = validate_spec_traceability(requirement_records, spec_content)

    ux_checks = {
        "flowchart_present": bool((ux_artifact or {}).get("flowchart")),
        "page_hierarchy_present": bool((ux_artifact or {}).get("pageHierarchy")),
    }
    result["ux_checks"] = ux_checks
    result["ok"] = result["ok"] and all(ux_checks.values()) if ux_artifact else result["ok"]
    return result


def run_plan_self_review(
    requirements: List[Dict[str, Any]],
    plan_content: str,
) -> Dict[str, Any]:
    requirement_records = build_requirements(requirements)
    tasks = parse_tasks_v2(plan_content)
    result = validate_plan_traceability(requirement_records, tasks)
    tasks_missing_verification = [task.id for task in tasks if not task.verification or not task.verification.commands]
    result["tasks_missing_verification"] = tasks_missing_verification
    result["ok"] = result["ok"] and not tasks_missing_verification
    return result


def run_doc_contract_review(
    cli_content: str,
    overview_doc_content: str,
    plan_template_content: str,
    other_doc_contents: List[str],
    existing_script_names: List[str],
) -> Dict[str, Any]:
    return validate_workflow_doc_contracts(
        cli_content,
        overview_doc_content,
        plan_template_content,
        other_doc_contents,
        existing_script_names,
    )


def main() -> int:
    import argparse

    parser = argparse.ArgumentParser(description="workflow self-review helper")
    sub = parser.add_subparsers(dest="command")

    p_spec = sub.add_parser("spec", help="运行 spec self-review")
    p_spec.add_argument("requirements_json")
    p_spec.add_argument("spec_file")
    p_spec.add_argument("--ux-json", default="{}")

    p_plan = sub.add_parser("plan", help="运行 plan self-review")
    p_plan.add_argument("requirements_json")
    p_plan.add_argument("plan_file")

    args = parser.parse_args()

    if args.command == "spec":
        with open(args.requirements_json, "r", encoding="utf-8") as f:
            requirements = json.load(f)
        with open(args.spec_file, "r", encoding="utf-8") as f:
            spec_content = f.read()
        print(json.dumps(run_spec_self_review(requirements, spec_content, json.loads(args.ux_json)), ensure_ascii=False))
        return 0

    if args.command == "plan":
        with open(args.requirements_json, "r", encoding="utf-8") as f:
            requirements = json.load(f)
        with open(args.plan_file, "r", encoding="utf-8") as f:
            plan_content = f.read()
        print(json.dumps(run_plan_self_review(requirements, plan_content), ensure_ascii=False))
        return 0

    parser.print_help()
    return 1


if __name__ == "__main__":
    sys.exit(main())

File: core/utils/workflow/self_review.js
#!/usr/bin/env node

const fs = require('fs')
const { validateWorkflowDocContracts } = require('./doc_contracts')
const { parseTasksV2 } = require('./task_parser')
const {
  validatePlanTraceability,
  validateSpecTraceability,
} = require('./traceability')

function buildRequirements(items = []) {
  return items.filter((item) => item.id).map((item) => ({
    id: item.id,
    summary: item.summary || '',
    scope_status: item.scope_status || 'in_scope',
    constraints: item.constraints || [],
    owner: item.owner || 'shared',
    exclusion_reason: item.exclusion_reason || null,
  }))
}

function runSpecSelfReview(requirements, specContent, uxArtifact = null) {
  const requirementRecords = buildRequirements(requirements)
  const result = validateSpecTraceability(requirementRecords, specContent)
  const uxChecks = {
    flowchart_present: Boolean((uxArtifact || {}).flowchart),
    page_hierarchy_present: Boolean((uxArtifact || {}).pageHierarchy),
  }
  result.ux_checks = uxChecks
  result.ok = uxArtifact ? result.ok && Object.values(uxChecks).every(Boolean) : result.ok
  return result
}

function runPlanSelfReview(requirements, planContent) {
  const requirementRecords = buildRequirements(requirements)
  const tasks = parseTasksV2(planContent)
  const result = validatePlanTraceability(requirementRecords, tasks)
  const tasksMissingVerification = tasks.filter((task) => !task.verification || !(task.verification.commands || []).length).map((task) => task.id)
  result.tasks_missing_verification = tasksMissingVerification
  result.ok = result.ok && tasksMissingVerification.length === 0
  return result
}

function runDocContractReview(cliContent, overviewDocContent, planTemplateContent, otherDocContents, existingScriptNames) {
  return validateWorkflowDocContracts(cliContent, overviewDocContent, planTemplateContent, otherDocContents, existingScriptNames)
}

function main() {
  const args = [...process.argv.slice(2)]
  const command = args.shift()
  if (command === 'spec') {
    const requirements = JSON.parse(fs.readFileSync(args[0], 'utf8'))
    const specContent = fs.readFileSync(args[1], 'utf8')
    const uxIndex = args.indexOf('--ux-json')
    const uxArtifact = uxIndex >= 0 ? JSON.parse(args[uxIndex + 1]) : {}
    process.stdout.write(`${JSON.stringify(runSpecSelfReview(requirements, specContent, uxArtifact))}\n`)
    return
  }
  if (command === 'plan') {
    const requirements = JSON.parse(fs.readFileSync(args[0], 'utf8'))
    const planContent = fs.readFileSync(args[1], 'utf8')
    process.stdout.write(`${JSON.stringify(runPlanSelfReview(requirements, planContent))}\n`)
    return
  }
  process.stderr.write('Usage: node self_review.js <spec|plan> ...\n')
  process.exitCode = 1
}

module.exports = {
  buildRequirements,
  runSpecSelfReview,
  runPlanSelfReview,
  runDocContractReview,
}

if (require.main === module) main()
